### PR TITLE
Fixed DMA validation test for new hw context

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
#### Problem solved by the commit
[CR-1159505](https://jira.xilinx.com/browse/CR-1159505) system crash during xrt stress test
-- The root cause of the issue is, as per the current design BO should be associated with a hw context. 
    If a BO is created without a context and there is a chance that the xclbin is replaced then Driver will through a error that BO is present while XCLBIN is replacing.  

-- This issue only happen for stress test. Single validation will not hit this issue.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
-- Generic issue of stress test. 
-- Hw context & Multislot design is introduced this issue. 

#### How problem was solved, alternative solutions (if any) and why they were rejected
-- Added hw context creation before allocating a BO. 
-- All legacy APIs need to update/remove. 

#### Risks (if any) associated the changes in the commit
-- N/A

#### What has been tested and how, request additional testing if necessary
xbutil validation  and stress test script 

#### Documentation impact (if any)
N/A

### CODE REVIEW

File: src/runtime_src/core/pcie/common/dmatest.h [hunk 1]
```diff
@@ -130,10 +131,13 @@ namespace xcldev {
             if (count > 0x40000)
                 count = 0x40000;
 
+	    xrt::hw_context::cfg_param_type cfg_param;
+	    mhwCtxHandle = mHandle->create_hw_context(mHandle->get_xclbin_uuid().get(), cfg_param, xrt::hw_context::access_mode::shared);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/common/dmatest.h:135]:
```suggestion
	    mhwCtxHandle = mHandle->create_hw_context(mHandle->get_xclbin_uuid().get(), {}, xrt::hw_context::access_mode::shared);
```

File: src/runtime_src/core/pcie/common/dmatest.h [hunk 2]
```diff
@@ -144,7 +148,9 @@ namespace xcldev {
         }
 
         ~DMARunner()
-        {}
+        {
+	    mBOList.clear();
```
Comment (stsoe) [inline src/runtime_src/core/pcie/common/dmatest.h:152]: Why?

**Review by stsoe — CHANGES_REQUESTED** [review]

**Comment by stsoe** [comment]
Also, please add PR description.

**Comment by saifuddin-xilinx** [comment]
> Also, please add PR description.

Sure, Actually I was just checking the pipeline. Hence, I have added the flag "Please do no Review". :)
I will fix the review comments as suggested. Thanks

File: src/runtime_src/core/pcie/common/dmatest.h (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/common/dmatest.h:152]: The BOs are getting freed after hw destructor getting called. 
Hence, after hw context destructed there are some BOs are still present.

File: src/runtime_src/core/pcie/common/dmatest.h [hunk 3]
```diff
@@ -130,10 +131,13 @@ namespace xcldev {
             if (count > 0x40000)
                 count = 0x40000;
 
+	    xrt::hw_context::cfg_param_type cfg_param;
+	    mhwCtxHandle = mHandle->create_hw_context(mHandle->get_xclbin_uuid().get(), cfg_param, xrt::hw_context::access_mode::shared);
+	    mFlags |= ((mFlags & ~XRT_BO_FLAGS_MEMIDX_MASK) | (mhwCtxHandle->get_slotidx() << XRT_BO_FLAGS_SLOTIDX_SHIFT));
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/common/dmatest.h]: Thanks. Fixed

File: src/runtime_src/core/pcie/common/dmatest.h (hunk 2 above)
Comment (stsoe) [reply src/runtime_src/core/pcie/common/dmatest.h:154]: But mBOList is an automatic object, it is destructed when DMARunner goes out of scope.  That's what I meant by "Why?"

Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/common/dmatest.h:154]: Buffers are keeping an information about hw context. If buffers are freed before hw context, then we are losing track. We need to make sure all buffers are freed up before hardware context goes out of scope.

Comment (stsoe) [reply src/runtime_src/core/pcie/common/dmatest.h:154]: Ok, got it, but order of destruction is opposite order of construction so it is implicitly correct anyway.  But I like being explicit about it, but that definitely requires a comment in the code before calling clear(), it is not at all obvious.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/pcie/common/dmatest.h (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/common/dmatest.h:156]: Added a comments for the same

**Comment by saifuddin-xilinx** [comment]
retest this please

**Pull request merged**